lib/sql.ts

```diff
@@ -0,0 +1,208 @@
+/**
+ * Async worker thread wrapper around SQLite, written to improve concurrent performance.
+ * @author mia-pi-git
+ */
+import {ProcessWrapper, ProcessManager} from './process-manager';
+import * as Sqlite from 'better-sqlite3';
+import * as child_process from 'child_process';
+import * as path from 'path';
+import {FS} from './fs';
+
+interface SQLOptions {
+	file: string;
+	/** file to import database functions from - this should be relative to this filename. */
+	extension?: string;
+}
+
+type DataType = unknown[] | Record<string, unknown>;
+
+export type DatabaseQuery = {
+	/** Prepare a statement - data is the statement. */
+	type: 'prepare', data: string,
+} | {
+	/** Get all lines from a statement. Data is the params, num is the statement number. */
+	type: 'all', data: DataType, num: number,
+} | {
+	/** Execute raw SQL in the database. */
+	type: "exec", data: string,
+} | {
+	/** Get one line from a prepared statement. */
+	type: 'get', data: DataType, num: number,
+} | {
+	/** Run a prepared statement. */
+	type: 'run', data: DataType, num: number,
+} | {
+	type: 'transaction', num: number, data: DataType,
+};
+
+export class DatabaseWrapper implements ProcessWrapper {
+	statements: Map<string, number>;
+	process: child_process.ChildProcess;
+	pendingRequests: ((data: string | DataType | null) => any)[];
+	constructor(options: SQLOptions) {
+		this.statements = new Map();
+		this.pendingRequests = [];
+		this.process = child_process.fork(__filename, [], {
+			env: options as AnyObject, cwd: path.resolve(__dirname, '..'),
+		});
+		this.listen();
+	}
+	getProcess() {
+		return this.process;
+	}
+	listen() {
+		this.process.on("message", (message: DataType) => {
+			const resolver = this.pendingRequests.shift();
+			if (resolver) return resolver(message);
+			throw new Error(`Database wrapper received a message, but there was no pending request.`);
+		});
+	}
+	release() {
+		this.statements.clear();
+		this.process.kill();
+		for (const resolver of this.pendingRequests.values()) {
+			resolver(null);
+		}
+		return Promise.resolve();
+	}
+	get load() {
+		return this.pendingRequests.length;
+	}
+	runFile(filename: string) {
+		const file = FS(filename);
+		if (!file.existsSync()) throw new Error(`File passed to runFile does not exist.`);
+		if (!filename.endsWith('.sql')) throw new Error(`File passed to runFile is not a .sql file.`);
+		const content = file.readSync();
+		void this.exec(content);
+	}
+	async prepare(statement: string) {
+		const cachedStatement = this.statements.get(statement);
+		if (cachedStatement) {
+			return cachedStatement;
+		}
+		const int = await this.query({type: 'prepare', data: statement});
+		if (typeof int === 'number') this.statements.set(statement, int);
+		return int;
+	}
+	all(statement: string | number, data: DataType = {}) {
```

Comment:
Can you make some sort of `Statement` type instead of having `string | number` everywhere?